src/coreclr/tools/Common/Pgo/PgoFormat.cs

```diff
@@ -0,0 +1,554 @@
+                        else
+                            lastDataValue = value;
+                        curSchema.DataLong = lastDataValue;
+                    }
+                    else
+                    {
+                        int dataIndex = curSchema.Count - dataCountToRead;
+                        switch(curSchema.InstrumentationKind & PgoInstrumentationKind.MarshalMask)
+                        {
+                            case PgoInstrumentationKind.FourByte:
+                                if (longsAreCompressed)
+                                    lastDataValue += value;
+                                else
+                                    lastDataValue = value;
+                                ((int[])curSchema.DataObject)[dataIndex] = checked((int)lastDataValue);
+                                break;
+
+                            case PgoInstrumentationKind.EightByte:
+                                if (longsAreCompressed)
+                                    lastDataValue += value;
+                                else
+                                    lastDataValue = value;
+                                ((long[])curSchema.DataObject)[dataIndex] = lastDataValue;
+                                break;
+
+                            case PgoInstrumentationKind.TypeHandle:
+                                if (longsAreCompressed)
+                                    lastTypeValue += value;
+                                else
+                                    lastTypeValue = value;
+                                ((TType[])curSchema.DataObject)[dataIndex] = dataProvider.TypeFromLong(lastTypeValue);
+                                break;
+                        }
+                    }
+                    dataCountToRead--;
+                    if (dataCountToRead == 0)
+                    {
+                        yield return curSchema;
+                        curSchema.DataLong = 0;
+                        curSchema.DataObject = null;
+                    }
+                    continue;
+                }
+
+                if (processingState == InstrumentationDataProcessingState.UpdateProcessMaskFlag)
+                {
+                    processingState = (InstrumentationDataProcessingState)value;
+                    continue;
+                }
+
+                if ((processingState & InstrumentationDataProcessingState.ILOffset) == InstrumentationDataProcessingState.ILOffset)
+                {
+                    if (longsAreCompressed)
+                        curSchema.ILOffset += checked((int)value);
+                    else
+                        curSchema.ILOffset = checked((int)value);
+
+                    processingState = processingState & ~InstrumentationDataProcessingState.ILOffset;
+                }
+                else if ((processingState & InstrumentationDataProcessingState.Type) == InstrumentationDataProcessingState.Type)
```

Comment:
Nit - it might make the code slightly easier to read if for single-bit masks we used just <code>!=0</code> instead of duplicating the bitmask constant.